### PR TITLE
LAB-367: Modernize the IDE page

### PR DESCRIPTION
**Pull request by RN35**
# Proposed Changes

- Migrated from create-react-app to Next.js 
- Added Material UI support 
- Removed the backend API code 
   - Next.js has capability to write APIs (https://nextjs.org/docs/api-routes/introduction)
- Removed the express.js based backend code execution code
   - This will be offloaded to Judge0 in a future JIRA (https://github.com/judge0/judge0)
- Created a basic layout ( NavBar, Drawer, Profile Menu) - All pages will inherit this (see: https://nextjs.org/docs/basic-features/layouts)
- Created the IDE page using Material UI components
   - Tabs to show (Problem description, Solution and past submissions) - These contain placeholder content for now
   - Integrated a code editor - Monaco (https://github.com/microsoft/monaco-editor)

# Jira Stories

- [LAB-367](https://oregonstate-innovationlab.atlassian.net/browse/LAB-367)

# Checklist

- [x] This has been tested
- [x] Self-documenting code
- [x] Code is linted

# Instructions to review
## How to Run
- Clone the repository: `git clone git@github.com:RN35/yacp.git`
- Change dir: `cd yacp`
- Checkout branch migrate-to-nextjs: `git checkout migrate-to-next-js`
- Install dependencies: `npm install`
- Spin up development server: `npm run dev`

## How to Validate
- Open `http://localhost:3000/problems` in any web browser
- This is just the UI, so just make sure everything renders correctly and there are no errors in the console 

# Screenshots 

### Drawer opens on the clicking the `hamburger menu`

image

### A dropdown menu opens on clicking the profile icon on the top-right of the navbar

image

### The IDE page

image

### CODE REVIEW

File: pages/_document.jsx [hunk 1]
```diff
@@ -0,0 +1,84 @@
+import * as React from "react";
+import Document, { Html, Head, Main, NextScript } from "next/document";
+import createEmotionServer from "@emotion/server/create-instance";
+import theme, { roboto } from "../src/theme";
+import createEmotionCache from "../src/createEmotionCache";
+
+export default class MyDocument extends Document {
+  render() {
+    return (
+      <Html lang="en" className={roboto.className}>
```
Comment (supreethavadhani) [inline pages/_document.jsx:10]: keep the tag casing to lower only <html> / <head>

File: pages/problems.jsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  return (
+    <Box sx={{ flexGrow: 1 }}>
+      <Grid container spacing={{ xs: 1, md: 1, lg: 1, xl: 1 }}>
+        <Grid
+          item
+          xs={12}
+          sm={12}
+          md={6}
+          key={1}
+          // sx={{ maxHeight: "100vh", overflowY: "scroll" }}
```
Comment (supreethavadhani) [inline pages/problems.jsx:42]: nitpick : rouge comment?

**Comment by ruangroc** [comment]
Looks good overall! Just some small suggestions:

Not sure what's going on in the top right corner but looks like a hidden scrollbar that doesn't move? Is there a way to clean that up a bit?
image

Could the output box be a fixed size or just slightly larger? (More padding maybe?) And could all the things on the right side of the screen be aligned to the same edge? Seems like there's some differences.
image

**Comment by supreethavadhani** [comment]
+1 on @ruangroc's comments

In the open dropdown menu it looks the profile icon is disappearing 
Screenshot 2023-02-24 at 9 52 31 PM

**Review by ruangroc — APPROVED** [review]

File: pages/problems.jsx (hunk 2 above)
Comment (RN35) [reply pages/problems.jsx]: Fixed

**Comment by RN35** [comment]
> Looks good overall! Just some small suggestions:
> 
> Not sure what's going on in the top right corner but looks like a hidden scrollbar that doesn't move? Is there a way to clean that up a bit? image
> 
> Could the output box be a fixed size or just slightly larger? (More padding maybe?) And could all the things on the right side of the screen be aligned to the same edge? Seems like there's some differences. image

@ruangroc thank you for taking the time to review it :)

the "hidden scrollbar you are referring it belongs to the editor itself. It is part of the Monaco editor. 
image

The output box currently expands based on the content. However this adds a scrollbar to the page. This is not the desired behavior. I have created a ticket to address this in the future https://oregonstate-innovationlab.atlassian.net/browse/LAB-375. 

Output box - I have added some padding and margin to make it look better. 
image

**Comment by RN35** [comment]
> +1 on @ruangroc's comments
> 
> In the open dropdown menu it looks the profile icon is disappearing Screenshot 2023-02-24 at 9 52 31 PM

@supreethavadhani thanks for reviewing this PR! 

Fixed :)
image

**Pull request merged**